DCOS-21232: Better text filtering

brandonc opened the pull request:
https://jira.mesosphere.com/browse/DCOS-21232

StringUtil#filterByString now sorts matches by relevance score. This change affects node health checks and universe packages. Here's how scoring is calculated:

- Search phrase are tokenized and each term is evaluated individually, with multiple scores summed together. So if you type "foo bar baz" a package named "foo baz" would appear before "foo" or "bar"
- Exact term matches are scored very very high
- Otherwise, longer terms appearing closer to the beginning of the string are scored higher than shorter terms closer to the end
- Additionally, universe packages are re-ranked to show certified packages first.

The "string" being searched is the concatenated values of "filterProperties"... For universe packages, that would be the name, description, and tags.

I got some ideas from fuse.js and Sifterjs, but ultimately chose an algorithm that felt best.

## Testing

1. Visit catalog
2. Observe exact matches like "beta-kubernetes" rank before more proximate packages like "kubernetes"
3. Observe search phrase "kaf" has more relevant ranking

1. Visit node details health tab
2. Observe that text filtering still works and ranks according to the rules outlined above

## Trade-offs

It is plausible that future callers of StringUtil#filterByString may not want reordering to take place.

## Screenshots

Before:

<img width="939" alt="screen shot 2018-09-20 at 4 36 54 pm" src="https://user-images.githubusercontent.com/174332/45850462-7f15a580-bcf3-11e8-8f25-f6a2e86e6b98.png">


After:

<img width="948" alt="screen shot 2018-09-20 at 4 34 25 pm" src="https://user-images.githubusercontent.com/174332/45850466-8472f000-bcf3-11e8-85e6-345f5ac46f4a.png">

GeorgiSTodorov commented:
![screenshot from 2018-09-21 10-31-43](https://user-images.githubusercontent.com/40791275/45866964-f2b6c180-bd8a-11e8-9296-88ac038d7ebe.png)

`beta-kubernetes` is 4th, even though it is returned as 2nd
![screenshot from 2018-09-21 10-41-19](https://user-images.githubusercontent.com/40791275/45867005-06622800-bd8b-11e8-8dfd-2a780dc0bb8a.png)

and it has the 2nd highest score
![screenshot from 2018-09-21 10-41-39](https://user-images.githubusercontent.com/40791275/45867016-0cf09f80-bd8b-11e8-90db-e802ff1736d6.png)

GeorgiSTodorov commented:
![screenshot from 2018-09-21 10-51-26](https://user-images.githubusercontent.com/40791275/45867509-4d9ce880-bd8c-11e8-8ca0-e88b237d7f76.png)

It works fine on master.

DanielMSchmidt commented:
Cassandra ordering also looks suboptimal 🤔 I wonder where we can draw the line for "good enough"
![bildschirmfoto 2018-09-21 um 15 37 34](https://user-images.githubusercontent.com/1337046/45884518-6ae5ad00-bdb4-11e8-9c5f-3dd790756188.png)

DanielMSchmidt commented on src/js/structs/__tests__/UniversePackagesList-test.js:
```diff
@@ -22,6 +22,25 @@ describe("UniversePackagesList", function() {
       expect(items[0].get("name")).toEqual("bar");
     });
 
+    it("sorts exact matches first", function() {
```
Would like to see some more tests for the items by text filter

brandonc commented:
@GeorgiSTodorov @DanielMSchmidt The scoring is doing it's job but the "certified" re-ranking is messing with the relevance sort of the community packages!

![screen shot 2018-09-21 at 10 54 11 am](https://user-images.githubusercontent.com/174332/45894827-c69c4000-bd8c-11e8-8c5e-39889b10e6b4.png)

brandonc commented:
@GeorgiSTodorov @DanielMSchmidt I modified the re-ranking for certified packages and that seemed to improve things. I also added some relevance-specific tests. Do you mind reevaluating this?

brandonc commented:
@DanielMSchmidt I extracted search stuff into a util and added typescript annotations. I'm not experienced with TS so don't hold back on the criticism.
